src/hotspot/share/runtime/handshake.cpp

```diff
@@ -465,60 +443,87 @@ bool HandshakeState::possibly_can_process_handshake() {
   }
 }
 
-bool HandshakeState::claim_handshake(bool is_direct) {
-  if (!_processing_sem.trywait()) {
+bool HandshakeState::claim_handshake() {
+  if (!_lock.try_lock()) {
     return false;
   }
-  if (has_specific_operation(is_direct)){
-    return true;
+  // Operations are added without lock and then the poll is armed.
+  // If all handshake operations for the handshakee are finished and someone
+  // just adds an operation we may see it here. But if the handshakee is not
+  // armed yet it is not safe to procced.
+  if (has_operation_for_processor()) {
+    if (SafepointMechanism::local_poll_armed(_handshakee)) {
+      return true;
+    }
   }
-  _processing_sem.signal();
+  _lock.unlock();
   return false;
 }
 
-HandshakeState::ProcessResult HandshakeState::try_process(HandshakeOperation* op) {
-  bool is_direct = op->is_direct();
+const char* executioner_name(Thread* current_thread, Thread* handshakee, bool current_is_requester) {
+  if (current_thread == handshakee) return "self(JavaThread)";
```

Comment:
I think we can remove this line since executioner_name() is only called by the handshaker.

Reply:
Yes, fixing.